contracts/R0_SpokePool.sol

```diff
@@ -0,0 +1,183 @@
+    error NotImplemented();
+    error InvalidSteelCommitment();
+
+    // All calls that have admin privileges must be fired from within the receiveL1State method that validates that
+    // the input data was published on L1 by the HubPool. This input data is then executed on this contract.
+    // This modifier sets the adminCallValidated variable so this condition can be checked in _requireAdminSender().
+    modifier validateInternalCalls() {
+        // Make sure adminCallValidated is set to True only once at beginning of the function, which prevents
+        // the function from being re-entered.
+        if (_adminCallValidated) {
+            revert AdminCallAlreadySet();
+        }
+
+        // This sets a variable indicating that we're now inside a validated call.
+        // Note: this is used by other methods to ensure that this call has been validated by this method and is not
+        // spoofed.
+        _adminCallValidated = true;
+
+        _;
+
+        // Reset adminCallValidated to false to disallow admin calls after this method exits.
+        _adminCallValidated = false;
+    }
+
+    /// @custom:oz-upgrades-unsafe-allow constructor
+    constructor(
+        address _steel,
+        address _verifier,
+        address _hubPoolStore,
+        address _wrappedNativeTokenAddress,
+        uint32 _depositQuoteTimeBuffer,
+        uint32 _fillDeadlineBuffer
+    ) SpokePool(_wrappedNativeTokenAddress, _depositQuoteTimeBuffer, _fillDeadlineBuffer, OFT_FEE_CAP) {
+        verifier = _verifier;
+        steel = _steel;
+        hubPoolStore = _hubPoolStore;
+    }
+
+    function initialize(
+        uint32 _initialDepositId,
+        address _crossDomainAdmin,
+        address _withdrawalRecipient
+    ) public initializer {
+        __SpokePool_init(_initialDepositId, _crossDomainAdmin, _withdrawalRecipient);
+    }
+
+    /**
+     * @notice This can be called by an EOA to relay events that the HubPool emits on L1.
+     * @param journalData The public data written by the guest program
+     * @param seal The encoded cryptographic proof (i.e. SNARK).
+     */
+    function receiveL1State(bytes calldata journalData, bytes calldata seal) external validateInternalCalls {
+        // Decode and validate the journal data
+        Journal memory journal = abi.decode(journalData, (Journal));
+        if (journal.contractAddress != hubPoolStore) {
+            revert NotHubPoolStore();
+        }
+        if (journal.eventParams_target != address(this)) {
+            revert NotTarget();
+        }
+        if (IHeliosSteelValidator(steel).validateCommitment(journal.commitment)) {
+            revert InvalidSteelCommitment();
+        }
+
+        // Verify the proof
+        bytes32 journalHash = sha256(journalData);
+        IRiscZeroVerifier(verifier).verify(seal, imageId, journalHash);
+
+        // Prevent replay attacks by using storage key which includes a nonce. The only way for someone to re-execute
```

Comment:
Sorry, are you saying the message includes a nonce? Or if not, how is the nonce included in the dataHash?

Reply:
https://github.com/across-protocol/contracts/pull/916/commits/74685ce8b2b9cdf3bef4faf69069cae5459cdf29